modules/navier_stokes/src/fvkernels/INSFVMomentumDiffusion.C

```diff
@@ -30,28 +33,60 @@ INSFVMomentumDiffusion::validParams()
                              "Switch that can select face interpolation method for the viscosity.");
 
   params.set<unsigned short>("ghost_layers") = 2;
+  params.addParam<bool>(
+      "complete_expansion",
+      false,
+      "Boolean parameter to use complete momentum expansion is the diffusion term.");
+  params.addCoupledVar("u", "The velocity in the x direction.");
+  params.addCoupledVar("v", "The velocity in the y direction.");
+  params.addCoupledVar("w", "The velocity in the z direction.");
   return params;
 }
 
 INSFVMomentumDiffusion::INSFVMomentumDiffusion(const InputParameters & params)
   : INSFVFluxKernel(params),
     _mu(getFunctor<ADReal>(NS::mu)),
-    _mu_interp_method(Moose::FV::selectInterpolationMethod(getParam<MooseEnum>("mu_interp_method")))
+    _mu_interp_method(
+        Moose::FV::selectInterpolationMethod(getParam<MooseEnum>("mu_interp_method"))),
+    _u_var(params.isParamValid("u")
+               ? dynamic_cast<const INSFVVelocityVariable *>(getFieldVar("u", 0))
+               : nullptr),
+    _v_var(params.isParamValid("v")
+               ? dynamic_cast<const INSFVVelocityVariable *>(getFieldVar("v", 0))
+               : nullptr),
+    _w_var(params.isParamValid("w")
+               ? dynamic_cast<const INSFVVelocityVariable *>(getFieldVar("w", 0))
+               : nullptr),
+    _complete_expansion(getParam<bool>("complete_expansion")),
+    _dim(_subproblem.mesh().dimension())
 {
   if ((_var.faceInterpolationMethod() == Moose::FV::InterpMethod::SkewCorrectedAverage) &&
       (_tid == 0))
     adjustRMGhostLayers(std::max((unsigned short)(3), _pars.get<unsigned short>("ghost_layers")));
+
+  if (_complete_expansion && !_u_var)
+    paramError("u", "The u velocity must be defined when 'complete_expansion=true'.");
+
+  if (_complete_expansion && _dim >= 2 && !_v_var)
+    paramError("v",
+               "The v velocity must be defined when 'complete_expansion=true'"
+               "and problem dimension is larger or equal to 2.");
+
+  if (_complete_expansion && _dim >= 3 && !_w_var)
+    paramError("w",
+               "The w velocity must be defined when 'complete_expansion=true'"
+               "and problem dimension is larger or equal to three.");
 }
 
 ADReal
 INSFVMomentumDiffusion::computeStrongResidual(const bool populate_a_coeffs)
 {
-  const auto state = determineState();
+  const Moose::StateArg state = determineState();
   const auto dudn = gradUDotNormal(state);
   ADReal face_mu;
 
   if (onBoundary(*_face_info))
-    face_mu = _mu(makeCDFace(*_face_info), state);
+    face_mu = std::max(_mu(makeCDFace(*_face_info), state), 0.0);
```

Comment:
We still should not accept negative viscosity values. This will destroy our convergence. What do you think of my change?

Reply:
I understand what the purpose of this change is. I will give the user the option to choose if they want to limit or not. I don't think this should be default. Not because I think that it is fine to have negative viscosity. I just don't like limiting things without the user's awareness. People can turn this limiting off and on. If we want to discard solutions which were obtained with negative viscosity values, we have to use the solution invalidity interface. Also, this should not be a problem if the turbulent k/epsilon are limited to be in the (0,inf) range. Negative values can come from the extrapolation for sure, especially if the gradient is large close to the boundary. You can disable the two-term expansion if that is the issue. But for now I'll let the user decide if we want to limit things during the iteration.